src/pcap/bosh.go

```diff
@@ -1,55 +1,414 @@
+		return fmt.Errorf("could not read bosh-director API response: %w", err)
+	}
+	err = json.Unmarshal(data, &apiResponse)
+	if err != nil {
+		return fmt.Errorf("could not parse bosh-director API response: %w", err)
+	}
+
+	br.uaaURLs = apiResponse.UserAuthentication.Options.URLs
+
+	br.logger.Info("connected to bosh-director", zap.Any("bosh-director", apiResponse))
+
+	return nil
+}
+
+func (br *BoshResolver) authenticate(authToken string) error {
+	allowed, err := br.verifyJWT(authToken)
+	if err != nil {
+		return fmt.Errorf("could not verify token: %w", err)
+	}
+
+	if !allowed {
+		return fmt.Errorf("token %s does not have the permissions or is not supported", authToken)
+	}
+
+	return nil
+}
+
+func (br *BoshResolver) getInstances(deployment string, authToken string) ([]BoshInstance, int, error) {
+	br.logger.Debug("checking token-permissions", zap.String("director-url", br.directorURL.String()), zap.String("deployment", deployment))
+	instancesURL, err := url.Parse(fmt.Sprintf("%s/deployments/%s/instances", br.directorURL, deployment))
+	if err != nil {
+		return nil, 0, err
+	}
+
+	req := &http.Request{
+		Method: http.MethodGet,
+		URL:    instancesURL,
+		Header: map[string][]string{
+			"Authorization": {"Bearer " + authToken},
+		},
+	}
+
+	res, err := br.client.Do(req)
+	if err != nil {
+		return nil, 0, fmt.Errorf("request to Bosh-director failed: %v", zap.Error(err))
+	}
+
+	defer res.Body.Close()
+
+	data, err := io.ReadAll(res.Body)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	if res.StatusCode != http.StatusOK {
+		return nil, res.StatusCode, fmt.Errorf("expected status code %d but got status code %d: %s", http.StatusOK, res.StatusCode, string(data))
+	}
+
+	var response []BoshInstance
+
+	err = json.Unmarshal(data, &response)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	return response, res.StatusCode, nil
+}
+
+// UaaKeyInfo holds the response of the UAA /token_keys endpoint.
+type UaaKeyInfo struct {
+	Kty   string `json:"kty"`
+	E     string `json:"e"`
+	Use   string `json:"use"`
+	Kid   string `json:"kid"`
+	Alg   string `json:"alg"`
+	Value string `json:"value"`
+	N     string `json:"n"`
+}
+
+// verifyJWT checks the JWT token in tokenString and ensures that it's valid and contains the neededScope as claim.
+// Validity is determined with the defaults, i.e.
+//   - validity time range
+//   - for RSA signed JWT that the RSA signature is consistent with the key provided by UAA
+//   - that there is a claim 'scope' that contains one entry that matches neededScope.
+//
+// Limitations: only RSA signed tokens are supported.
+//
+// returns a boolean that confirms that the token is valid, from a valid issuer and has the needed scope,
+// and an error in case anything went wrong while verifying the token and its scopes.
+func (br *BoshResolver) verifyJWT(tokenString string) (bool, error) {
+	token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
+		if jku, ok := token.Header["jku"]; ok {
```

Comment:
To get the function more readable, it is the better way to check if ok is false and return an error. Like
`jku, ok := token.Header["jku"]
  if !ok {
	return nil, fmt.Errorf("header 'jku' missing from token, cannot verify signature")
  }
  jkuURL, err := url.Parse(jku.(string))`

Reply:
Yes, fair.